### ISSUE TITLE
Dynamically create and destroy windows that share contexts with each other

### ISSUE THREAD

**Issue by duncanburke** [tobbebex/GPipe-Core#18]
I've been working on SDL bindings for GPipe based on the existing GLFW bindings and have some questions about support for shared contexts and the rationale behind the relevant design decisions.

As I understand, there are two primary use cases for shared contexts: firstly, when rendering to multiple windows to allow the individual contexts bound to each window to share resources; secondly to have a background thread with a shared context dedicated to loading assets.

As far as I can tell, despite nominally implementing support for shared contexts, it cannot be practically used for either of the mentioned use cases. Either of these use cases would require the ability to create two shared contexts and interleave the sending of commands to them, generally these contexts would be created at initialisation and persist throughout the lifetime of the program.

From the signature of runSharedContextT a child context is run wholly inside its parent context. At the completion of runSharedContextT the child context is destroyed and the flow of execution is once again in the parent context.

``` haskell
runSharedContextT :: ContextFormat c ds -> ContextT w os (ContextFormat c ds) (ContextT w os f m) a -> ContextT w os f m a
```

As ContextT is a monad transformer it may be expected that one could use `lift` to send commands to the parent context from the child action. However, the implementation in GPipe-GLFW uses the parent thread and communication channel for the child; the thread switches to the child context when in the child action and only switches back to the parent context when returning. With this model even if it were possible to send commands to the parent context from within the child action it would require interleaved switching between contexts on a single thread, which I believe would be quite expensive.

I believe that for shared contexts to be useful each context must have a dedicated thread and channel. It must be possible to send commands concurrently in both contexts. I also believe that safe communication between contexts would require protection by pipeline fences with SYNC_GPU_COMMANDS_COMPLETE, though I do not fully appreciate the ramifications this would have on the use of the `os` type parameter to ensure safety.

As a minor additional issue, newSharedContext implicitly uses whatever configuration was initially used to create the ContextHandle. In the example of rendering to multiple windows simultaneously it would be useful to be able to specify that the window of the child context be created at a different location to the parent. This is not critical as it is possible to move a window after creation, however this may not be true in general, especially for fullscreen windows.

**Comment by tobbebex** [comment]
" it would require interleaved switching between contexts on a single thread, which I believe would be quite expensive.

I believe that for shared contexts to be useful each context must have a dedicated thread and channel."

That assumption is actually false. I've worked quite extensively with multiple shared OpenGl contexts in the past, and have seen that it is actually less efficient to try to run two shared contexts concurrently from multiple threads, than to interleave them on a single thread (the driver is internally using a critical section for the shared contexts). The exception to this is dma-bound operations like texture and buffer writes and reads, that admittedly GPipe is not currently optimized for (your second use case). When running multiple contextes from a single thread you don't need to sync them using SYNC_GPU_COMMANDS_COMPLETE. Here are some other sources for you on this topic:
http://blog.gvnott.com/some-usefull-facts-about-multipul-opengl-contexts/
http://www.equalizergraphics.com/documentation/parallelOpenGLFAQ.html

As for the first use case, I can't really see what the problem is? Are you talking about the fact that the outer monad has to outlive the inner one? This is kind of a general problem for OpenGl shared contexts, and the solution is usually to have an outer windowless context and create shared contexts for each window. In GPipe, those windows would have to be wrapped though, meaning that you cannot create window A, then B, then close A before B. Having a forkContextT method (working just like the normal fork but in ContexT instead of IO) would fix this. I'll add that to my list :)

Your point about newSharedContext is perfectly valid. At the time this interface was created GPipe-GLFW didn't have a way to take parameters for the first context either so this design flaw passed by unnoticed. Would you mind creating a separate bug specifically for this?

**Comment by duncanburke** [comment]
Okay, I've created a [new issue](https://github.com/tobbebex/GPipe-Core/issues/19) for the newSharedContext problem and a [corresponding issue](https://github.com/plredmond/GPipe-GLFW/issues/10) in GPipe-GLFW.

**Comment by tobbebex** [comment]
I have created a ThreadedContexts branch where I added a forkSharedContextT. 

However, I'm having trouble realizing the backend in GPipe-GLFW. The problem is that most GLFW commands needs to be run from the main thread (including pollEvents and destroyWindow) so the threaded context would need to communicate with the originating thread, that in turn would need to poll that channel somehow (e.g. when swapping its own buffers, but what if its a windowless context?). Also, the threaded context couldn't outlive the originating threads context, so it would either have to block waiting for threaded shared contexts, or forcibly kill those.

So I am feeling a bit reluctant to this design now. What exactly are your requirements for shared contexts? Would it be ok to have the above mentioned limitations? 

Also inviting @plredmond to chip in any thoughts.

**Comment by plredmond** [comment]
The choice of using GLFW was largely based on a desire not to be limited by GLUT's feature set. GLFW supports a few heavier use-cases than GLUT, but is by no means a solution to all uses of GPipe/OpenGL. I welcome the creation of new context-handling backends, such as SDL @duncanburke.

With the currently defined GPipe [ContextHandle](https://github.com/tobbebex/GPipe-Core/blob/master/GPipe-Core/src/Graphics/GPipe/Internal/Context.hs#L55) API, there's no place in the `newSharedContext` function signature for backend-specific arguments to instruct GLFW as to how to construct the new context.

From `Graphics.GPipe.Internal.Context.{ContextFactory,ContextHandle}`:
```haskell
type ContextFactory c ds w = ContextFormat c ds -> IO (ContextHandle w)

...
    newSharedContext :: forall c ds. ContextFormat c ds -> IO (ContextHandle w),
```

If we can work together to come up with a backend-agnostic way of specifying capabilities, that would solve one of the issues you mentioned, @duncanburke.

I recognize that the larger issue @duncanburke is raising refers to how the current threading model may impact performance, but I have to admit I'm a beginner in graphics coding (though @tobbebex is experienced) and I don't know the best approach to juggling threads & current-contexts in GLFW. Suggestions are appreciated, and I'm happy to re-work GPipe-GLFW with your guidance.

**Comment by tobbebex** [comment]
Having cropped off issue #19, and dismissed the performance concern of using sharing a single thread for shared contexts, let's refine this issue down to this story:

I cannot as a user of GPipe 
* dynamically create and destroy windows that share contexts with each other.
* operate on a shared object space from two threads concurrently

Both of these will require some big changes to the Context API of GPipe. 

For the first bullet, I am envisioning making ContextT just represent the "object space" and have separate "windows" objects that you can create and destroy at will. 

I don't have a solution for the second bullet yet though. "Something" on the main thread will have to outlive the other threads for restrictions on how winproc is used in GLFW and others, but I don't want to go as far as GLUT went and take over the main thread completely with an ubiquitous mainLoop. Something along the line of [rust's scoped threads](https://doc.rust-lang.org/std/thread/fn.scoped.html) could be a way forward...

**Comment by plredmond** [comment]
**TLDR idea:** GPipe should hold a thread internally that has the object space but no window. Operations on the object space should use RPC to that thread. All other operations can operate on current context  through `ContextT`. No more `glfwMakeContextCurrent`.

---

> * operate on a shared object space from two threads concurrently
>
> I don't have a solution for the second bullet yet though. "Something" on the main thread will have to outlive the other threads...

I don't know whether "operate on a shared object space from two threads concurrently" is disallowed by our APIs or is simply disallowed in the backends [currently only GLFW] but if it's just a problem in the API, then your solution seems good. See below.

> I am envisioning making ContextT just represent the "object space" and have separate "windows" objects that you can create and destroy at will.

When some application-thread does `runContextT`, GPipe could create & hold its own OS thread internally to construct & maintain a windowless context and associated object space. Following that, using either of the mechanisms #19 `SharedContextFactory` or https://github.com/tobbebex/GPipe-Core/commit/c4bbb0e8129c382af4c7108f08f3ab5ba576854c `forkSharedContextT`, application code can create windowed contexts from within the `ContextT`.

The originating application-thread then is irrelevant and could terminate. There's no GLUT-style mainloop call in application code, but GPipe would be responsible for shutting down its internal thread when the last window object is destroyed.

Following this,

* Operations on the object space could be RPC'd to GPipe's internal context-originating thread.
* Operations on the application's current context can take place in that application thread.

If my understanding is correct, this would avoid the need to juggle contexts with `glfwMakeContextCurrent`, which has made the GPipe-GLFW code rather complex.

**Comment by tobbebex** [comment]
"operate on a shared object space from two threads concurrently" is kind of disallowed by both: GPipe doesnt have a fork-function, and at least GLFW (but probably other WMs) only allows some WM-commands (such as pollEvents) to be issued from the main thread. In addition to that, as I wrote in my first comment, issuing GL-commands concurrently from multiple threads in shared contexts are really bad for performance, with the exception for some DMA-bound operations like pixel transfers.

GPipe is already (or actually GPipe-GLFW is) creating one internal thread for each object space. The problem are those WM-commands that has to be run on the main thread. So then maybe runContextT should steal the main thread (it already requires to be run on that) and run a RPC-pump there, and instead create an additional thread in which the contextT-code is run (which will rpc any WM commands to the main thread and any GL commands to the respective object space thread).

By the way, GPipe is already quite conservative with making glfwMakeContextCurrent calls: It only does it when the GL-call is specific to the window (i.e. drawing to the default framebuffer or any operations on FBOs or VAOs, which are not shared even between shared contexts), while any other operation (like writing to a buffer or drawing to a texture) will go with whatever (shared) context was currently bound on the GL-thread. Note that non-shared object spaces are using separate GL-threads (that doesn't have any performance implications either; the driver only seems to implicitly lock as heavily for shared contexts).

**Comment by tobbebex** [comment]
Cropped off the threading part of this to #40.

**Comment by tobbebex** [comment]
Closing this as implemented with 32ef0ec6b1208ec53ab163976e40bf9b54c41ab5 and to be released in 2.2.

**Comment by tobbebex** [comment]
To summarize what was fixed in this ticket:
In 2.2 you are now using a single ContextT monad transformer layer and instead create window objects using newWindow and delete them with deleteWindow. These windows are then sent into shaders in the Shader environment (or passed in via the closure when the shader is created).
